### DISCUSSION TITLE
Testing multiple operating systems (os, windows) with golangci-lint in CI

### DISCUSSION THREAD

**Discussion by leap-pea** [golangci/golangci-lint#5159] [Q&A] [answered]
Hey,
what's the best way to run `golangci-lint run` for all OSs in the CI? Is there an option? Or do I need multiple invocations with `GOOS=windows`, etc?

**Comment by ldez** [comment, marked as answer]
Hi,

There is no option, golangci-lint relies on elements it has been compiled for.

Golangci-lint is not Go, using `GOOS`/`GOARCH` will not work as expected because of my previous explanation.


You can see examples inside the golangci-lint GitHub action: https://github.com/golangci/golangci-lint-action?tab=readme-ov-file#how-to-use

**Reply by wxiaoguang** [reply]
It does work

```
GOOS=linux golangci-lint run --build-tags=linux
GOOS=windows golangci-lint run --build-tags=windows
```

**Comment by wxiaoguang** [comment]
@ldez 

I think the answer is not true.

golangci-lint uses Golang compiler (parser), and it supports "--build-tags"

When it lints, there is nothing related to "elements it has been compiled for"

Just like you can cross-compile a Go program on Linux for Windows, you can also use golangci-lint to lint "windows" code on Linux, as long as you can make the "Go compiler" have correct tags, the real example is:

```
# this perfectly works on Linux, just like `GOOS=windows go build` the source code
GOOS=windows golangci-lint run --build-tags=windows
```

**Reply by ldez** [reply]
This is just wrong because golangci-lint is not Go.

You are confusing golangci-lint compilation and your program compilation.

golangci-lint is compiled for an OS and so code related to the other OS are not compiled inside golangci-lint.

To be super crystal clear about that: if a file **inside golangci-lint** has a build tag for an other OS, it will not be compiled **inside golangci-lint**.

This is the same thing for every program compiled for an OS.

<details>
<summary>main.go</summary>

```go
package main

import "fmt"

func main() {
	fmt.Println("UNIX")
}

```

</details>

<details>
<summary>main_windows.go</summary>

```go
//go:build windows

package main

import "fmt"

func main() {
	fmt.Println("WINDOWS")
}
```

</details>

```console
$ go build .                                                    
$ GOOS=windows ./sandbox                                      
UNIX
```

As I already said, you could run golangci-lint with `GOOS`/`GOARCH`, it may work in some cases, but if any element compiled inside golangci-lint relies on a build tags it will not work as expected.
Using `GOOS`/`GOARCH` is not doing what you think because golangci-lint is not Go.

**Reply by wxiaoguang** [reply]
See this:

<details><summary>prepare a "testlint" package</summary>

```console
$ cat testlint/demo.go
package testlint

func unusedDefault() {
}

$ cat testlint/demo_tag_win.go
//go:build windows

package testlint

func unusedTagWindows() {
}

$ cat testlint/demo_os_windows.go
package testlint

func unusedOsWindows() {
}
```

</details>

<details><summary>default "golangci-lint"</summary>

```console
$ golangci-lint run testlint
testlint/demo.go:3:6: func unusedDefault is unused (unused)
func unusedDefault() {
     ^
1 issues:
* unused: 1
```

</details>

<details><summary>cross-platform "golangci-lint"</summary>

```console
$ GOOS=windows golangci-lint run --build-tags=windows testlint
testlint/demo.go:3:6: func unusedDefault is unused (unused)
func unusedDefault() {
     ^
testlint/demo_os_windows.go:3:6: func unusedOsWindows is unused (unused)
func unusedOsWindows() {
     ^
testlint/demo_tag_win.go:5:6: func unusedTagWindows is unused (unused)
func unusedTagWindows() {
     ^
3 issues:
* unused: 3
```

</details>

**Reply by wxiaoguang** [reply]
> golangci-lint is compiled for an OS and so code related to the other OS are not compiled inside golangci-lint.

I think it is unrelated to the lint logic.

Lint only needs the project's source code and the related dependencies (including Golang std packages, go.mod packages).

Indeed, none of the project's source code is ever complied into golanglint-ci, golanglint-ci can still lint the whole project. So the "lint" is not related to OS, cross-OS code doesn't need to be complied into golanglint-ci

**Reply by ldez** [reply]
You don't want to understand, sorry but I cannot do more.
You can continue to think you understand even if you are off-topic.
But don't spread misinformation.

**Reply by wxiaoguang** [reply]
> This is just wrong because golangci-lint is not Go.
> You are confusing golangci-lint compilation and your program compilation.
> golangci-lint is compiled for an OS and so code related to the other OS are not compiled inside golangci-lint.

Well, maybe there were some misunderstanding of the discussion (the purpose of the "cross-platform" lint)

This discussion is for "use pre-compiled golangci-lint binary to lint the target source code for another platform"

The example I shown above is:

* use pre-compiled "golangci-lint" binary (macOS / Linux)
* it can lint the target source code for Windows platform (cross-lint), as long as the tags are right.

ps: you can assume that I have enough knowledge (20+ years) about various programming languages, CPU architectures and cross-compiling, very experienced in Golang and CGO/C/C++, etc 😄 

I cannot do more either, as long as it does work

**Reply by wxiaoguang** [reply]
> You don't want to understand, sorry but I cannot do more. You can continue to think you understand even if you are off topic. But don't spread misinformation.

At least, show a bad case for my example?

**Reply by ldez** [reply]
What do you don't understand in this sentence:
> As I already said, you could run golangci-lint with `GOOS`/`GOARCH`, it may work in some cases, but if any element **compiled inside golangci-lint** relies on a build tags it will not work as expected.
Using `GOOS`/`GOARCH` is not doing what you think because golangci-lint is not Go.

**Reply by wxiaoguang** [reply]
> but if any element compiled inside golangci-lint relies on a build tags it will not work as expected.

I think I have been pretty clear: use the pre-compiled golangci-lint, so it just works if "no element is needed to be complied inside golangci-lint", right?

If people need their own element to be complied by golangci-lint, they can't do cross-platform lint (just like CGO without gcc)

But for most people who only use the pre-built golangci-lint binary, it works perfectly, right?

**Reply by ldez** [reply]
You don't understand, you are totally off-topic: I don't say that golangci-lint needs to be compiled with the code of the project to analyze.

I said: if a linter, for example, is using build tags in its own code (this is not related to the project to lint), this code will not be inside golangci-lint because golangci-lint is compiled for an OS.

**Reply by wxiaoguang** [reply]
> I said: if a linter, for example, is using build tags in its own code (this is not related to the project to lint), this code will not be inside golangci-lint because golangci-lint is compiled for an OS.

I think I have been quite clear: **use pre-complieted golanglint-ci binary**, nothing need to compile, everything works, right?

Users can even `golanglint-ci custom` (with no GOOS pollution) to pre-build a binary with custom plugins, and then run it to lint  other platforms, is it right?

**Reply by ldez** [reply]
THIS IS NOT RELATED TO USING PRE-COMPILED BINARY.

Please read again my previous post.

**Reply by wxiaoguang** [reply]
The user is JUST ASKING FOR "what's the best way to run golangci-lint run for all OSs in the CI? Is there an option? Or do I need multiple invocations with GOOS=windows, etc?"


So, pre-compile a binary, then it does run for all OSs in the CI, no need to use matrix, is it right?

Which point is off-topic?

**Reply by ldez** [reply]
> So, pre-compile a binary, then it does run for all OSs in the CI, no need to use matrix, is it right?

No

> Which point is off-topic?

I give up, your are deep into a wrong interpretation of what I'm saying.

I repeated, reworded several times, provided examples, now, feel free to think what you want, I did my best to explain, sorry to not be able to do better.

**Reply by wxiaoguang** [reply]
> > So, pre-compile a binary, then it does run for all OSs in the CI, no need to use matrix, is it right?
> 
> No

There is just no bad case from you, then you just said "provided examples".

You said "THIS IS NOT RELATED TO USING PRE-COMPILED BINARY". Any real example that the pre-compiled binary doesn't work? Just like Linus said: talking is cheap, show me the code.

**Reply by wxiaoguang** [reply]
> What do you don't understand in this sentence:
> 
> > As I already said, you could run golangci-lint with `GOOS`/`GOARCH`, it may work in some cases, but if any element **compiled inside golangci-lint** relies on a build tags it will not work as expected.

Could you show me a real such "element"? I'd like to learn the details to see what indeed it is doing, or maybe its design was wrong.
